### ISSUE TITLE
Incremental Document updates

### ISSUE THREAD

**Issue by nihaopaul** [Hopding/pdf-lib#816]
can PDF-LIB support incremental updates? i loaded a document that was signed with incremental updates and then saved it again and noticed all of the incremental updates

startxref
n..
%%EOF

startxref
n..
%%EOF

startxref
n..
%%EOF

had been collapsed into one at the end, i was testing if pdf-lib could be used to build PAdes-LTV PDF's, the only part not found is the incremental updates but i've been amazed by this library so far. the only other way is that i manually append to the document but that gets messy on a double signature.

**Comment by therpobinski** [comment]
Hi, it is the same problem I have.
Have you found a solution?
I have a document in which I add a text before signing it. But adding the text with `pdf-lib` does not respect the incremental signature structure.
Before adding the second stamp:
```
%PDF-1.7
  <<content>>
  <<
    /Type /Sig
    /Filter /Adobe.PPKLite
    /SubFilter /adbe.pkcs7.detached
    /ByteRange [ 0 xx xxxx xxx ]
    /Contents <xxxxxxxxxxxxxxx>
  startxref
  xxxx
%%EOF
```
After adding the text:
```
%PDF-1.7
  <<content>>
  <<
    /Type /Sig
    /Filter /Adobe.PPKLite
    /SubFilter /adbe.pkcs7.detached
    /ByteRange [ 0 xx xxxx xxx ]
    /Contents <xxxxxxxxxxxxxxx>

  <<who adds the text content using pdf-lib>>

  startxref
  xxxx
%%EOF
```
But, the correct thing would be:
```
%PDF-1.7
  <<content>>
  <<
    /Type /Sig
    /Filter /Adobe.PPKLite
    /SubFilter /adbe.pkcs7.detached
    /ByteRange [ 0 xx xxxx xxx ]
    /Contents <xxxxxxxxxxxxxxx>
  startxref
  xxxx
%%EOF

<<who adds the text content using pdf-lib>>

  startxref
  xxxx
%%EOF
```

Do you have any ideas to correct this problem or is there no solution in this library?

**Comment by aashishvanand** [comment]
Hi, Any updates on this ?

**Comment by Hopding** [comment]
Not yet, but it's on the roadmap: https://github.com/Hopding/pdf-lib/discussions/998.

**Comment by aallvi** [comment]
hi, i am trying to do the same thing that your , how you get a solution? i need tu put image in a pdf that already have a signature.

 pdf-lib accept now the incremental saves and updates? you use another library? any help is appreciate.

Thanks

**Comment by PatrikAarnivaara** [comment]
> can PDF-LIB support incremental updates? i loaded a document that was signed with incremental updates and then saved it again and noticed all of the incremental updates
> 
> startxref n.. %%EOF
> 
> startxref n.. %%EOF
> 
> startxref n.. %%EOF
> 
> had been collapsed into one at the end, i was testing if pdf-lib could be used to build PAdes-LTV PDF's, the only part not found is the incremental updates but i've been amazed by this library so far. the only other way is that i manually append to the document but that gets messy on a double signature.

Any luck finding a solution in javascript?

**Comment by aallvi** [comment]
Hi patrik, i dont found a solution in javascript, i have to change this part of my aplication to python, with a library called endesive, is easy to use, i recommended to you to use

**Comment by PatrikAarnivaara** [comment]
Thanks, I also tried pymupdf. Well hummusjs (not maintained any longer) does incremental updates, I tried it and it works.

**Comment by adnsistemas** [comment]
Hello, there is a fork of the project that already implements incremental updates (https://github.com/remdra/pdf-lib-incremental-save), and I've used it to add signatures to PDFs without problems.
The only thing about this fork is that I think that doesn't follow a reasonable approach to incremental update, as you have to manually take a "snapshot" and register anything not new that has to be "incremented".
My idea is to take that fork and use it to make a fork that allows you to load a PDF for incremental update and automatically saves it with the right update, instead of a full rewrite. So IncrementalUpdate will be an option on document loading (no sense for new documents).
I'm willing to do the "merge" if you are going to eventually merge the PullRequest into the project.
I'll wait for your confirmation before starting to work on this, as I've seen you have a lot of requirements on collaborations (maybe not.. but I haven't done many collaborations..)
Thanks.
David.

**Comment by SamiSaves** [comment]
Hey @adnsistemas, there is an active fork of this package and they seem to be active with these kinds of requests and proposals. Could you consider opening an issue regarding this into that repository?

The updated fork can be found here: https://github.com/cantoo-scribe/pdf-lib

**Comment by adnsistemas** [comment]
I'm actually merging that fork into the fork I made from remdra.
Once I finish the merge will make a pull-request into that project.
Thanks.

**Comment by MrFerrys** [comment]
Hi folks,
I finally found the solution by using this project: [zgapdfsigner](https://github.com/zboris12/zgapdfsigner).
It uses a few dependencies such as pdf-lib, node-forge, and others.

Zgapdfsigner is a browser-friendly project that you can build in plain JavaScript.
To get it working in a browser with TSA and LTV signatures, you need to disable CORS.
You can do this by running your browser with the --disable-web-security flag, or simply by installing the Allow CORS extension (available for all major browsers):
[Allow Cors](https://chromewebstore.google.com/detail/allow-cors-access-control/lhobafahddgcelffkeicbaginigeejlf?hl=es).

Special thanks to remdra for helping me out.

**Comment by adnsistemas** [comment]
Hi, I forgot to let you know that @adnsistemas/pdf-lib package implements incremental update in a simpler way, based on remdra's fork. This fork is published as an NPM package and has updated documentation about its usage.
The changes I made are being merge into @cantoo-scribe/pdf-lib, so I guess will be available on that package too in the future.
Cheers.
